### ISSUE TITLE
Setting up HttpSys with Base Path

### ISSUE THREAD

**Issue by dannyfhalpotia** [aspnet/AspNetCore#13028]
I've tried out many different ways of setting the base path when using HttpSys but none seems to work once it gets deployed to a windows server. The funny thing is that my app is getting hit because the middleware is getting triggered. So https://localhost:30310/SomeBasePath (localhost on the server) will give a API IS RUNNING response and the same with app.Run for 404s.

The issue seems to be that ALL paths to my controllers resolve to a 404 and in the logs I get:

> Microsoft.AspNetCore.Builder.RouterMiddleware|Request did not match any routes

However, on my local machine it works fine and the route is matched. I also have the base path set up locally but one main difference is that I'm just using HTTP locally. I don't think the scheme is the problem as otherwise I wouldn't get to my application at all.
Some of my code:

**Program.cs CreateWebHostBuilder()**

```
public static IWebHost CreateWebHostBuilder(string[] args)
{
    IHostingEnvironment env = null;

    var builder =
        WebHost.CreateDefaultBuilder(args)
         .UseStartup<Startup>()
         .ConfigureAppConfiguration((hostingContext, config) =>
         {
             env = hostingContext.HostingEnvironment;
         })
         .UseHttpSys(options =>
         {
              options.Authentication.AllowAnonymous = true;
              if (!env.IsDevelopment())
              {
                  //options.UrlPrefixes.Add("https://*:30310/SomeBasePath/");
                  options.UrlPrefixes.Add("https://*:30310/");
              }
              else
              {
                  options.UrlPrefixes.Add("http://localhost:5000/SomeBasePath/"); //works fine
              }
         })
         .ConfigureLogging(b =>
         {
             //ommitted
         })
         .UseNLog()
         .Build();
    
    return builder;
}
```

**Startup.cs**

```
public void ConfigureServices(IServiceCollection services)
{
    //IoC and DB context init ommitted
    
    services.AddCors(c => 
    {
        c.AddPolicy("AllowOrigin", options => 
               options.AllowAnyOrigin().AllowAnyHeader().AllowAnyMethod());
    });

    services.AddMvc();
}

public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
    app.UsePathBase(new PathString("/SomeBasePath"));
    app.UseCors("AllowOrigin");

    //Conventional routing doesnt work either
    //app.UseMvc(routes =>
    //{
    //    routes.MapRoute(
    //    name: "default",
    //     template: "{controller=Home}/{action=Index}/{id?}");
   //});

   app.UseMvc();

    app.Use((context, next) =>
    {
        if (context.Request.Path.Value == "" || context.Request.Path.Value == "/")
        {
            context.Response.ContentType = "text/plain";
            return context.Response.WriteAsync("API IS RUNNING");
        }
        return next.Invoke();
    });

    app.Run(context =>
    {
        context.Response.StatusCode = 404;
        context.Response.ContentType = "application/json";

        _logger.Debug("Request Method: {METHOD}", context.Request.Method);
        _logger.Debug("Request Scheme: {SCHEME}", context.Request.Scheme);
        _logger.Debug("Request Path: {PATH}", context.Request.Path);
        _logger.Debug("Request Path Base: {PATH BASE}", context.Request.PathBase);

        // Headers
        foreach (var header in context.Request.Headers)
        {
             _logger.Debug("Header: {KEY}: {VALUE}", header.Key, header.Value);
        }
        // Connection: RemoteIp
        _logger.Debug("Request RemoteIp: {REMOTE_IP_ADDRESS}",
        context.Connection.RemoteIpAddress);
       
        return context.Response.WriteAsync("{ \"message\": \"Not found!!! \" }");
    });
```

**Simple controller:**

```
    [Route("api/system")]
    [ApiController]
    public class SystemController : ControllerBase
    {
        
        [HttpGet]
        [ProducesResponseType(StatusCodes.Status200OK)]
        public IActionResult Get()
        {
            return Ok("Hello World");
        }
    }
```

So, a path to the above through GET https://localhost:30310/SomeBasePath/api/system on the server gives a 404 and logs that no routes were matched. In addition the logs produced in my 404 middleware tells me that request path is "api/system" and that the path base is "SomeBasePath", so it seems that it's doing the right splitting into path and path base but still not able to match the path against any controller which is super weird.

I've tried to add the base path to the UrlPrefix in Program.cs and leave out the app.UsePathBase but with no luck. Any idea where I'm going wrong?

I should also add that I'm using a gateway normally to call my API through that publicly, and I'm getting the exact same issue when going through the gateway, so I've just resorted to testing locally on the server to rule out any potential gateway issues.

**Comment by Tratcher** [comment]
This line is redundant when you've defined the base path at the server level.
`app.UsePathBase(new PathString("/SomeBasePath"));` 

Please share the full logs from the good and bad repros.

**Comment by dannyfhalpotia** [comment]
Hi @Tratcher 

I did also try without the `app.UsePathBase` and have the base path in the `UrlPrefix` instead with the same problems persisting. Anyway, some logs:

Localhost on local machine where everything is working as it should:

```
warn: Microsoft.AspNetCore.Server.HttpSys.MessagePump[0]
      Overriding address(es) 'https://localhost:5001/, http://localhost:5000/'. Binding to endpoints added to UrlPrefixes instead.
info: Microsoft.AspNetCore.Server.HttpSys.HttpSysListener[0]
      Start
info: Microsoft.AspNetCore.Server.HttpSys.HttpSysListener[0]
      Listening on prefix: http://localhost:5000/
dbug: Microsoft.AspNetCore.Hosting.Internal.WebHost[4]
      Hosting started
dbug: Microsoft.AspNetCore.Hosting.Internal.WebHost[0]
      Loaded hosting startup assembly [REMOVED]
Hosting environment: Development
Content root path: [REMOVED]
Now listening on: http://localhost:5000/
Application started. Press Ctrl+C to shut down.
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:5000/somebasepath/api/system
dbug: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[0]
      Wildcard detected, all requests with hosts will be allowed.
trce: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[0]
      All hosts are allowed.
dbug: Microsoft.AspNetCore.Routing.Tree.TreeRouter[1]
      Request successfully matched the route with name '(null)' and template 'api/system'
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Route matched with {action = "Get", controller = "System"}. Executing controller action with signature Microsoft.AspNetCore.Mvc.IActionResult Get() on controller [REMOVED].Controllers.SystemController ([REMOVED]).
dbug: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Execution plan of authorization filters (in the following order): None
dbug: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Execution plan of resource filters (in the following order): Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter
dbug: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Execution plan of action filters (in the following order): Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter (Order: -3000), Microsoft.AspNetCore.Mvc.Infrastructure.ModelStateInvalidFilter (Order: -2000)
dbug: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Execution plan of exception filters (in the following order): None
dbug: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Execution plan of result filters (in the following order): Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Resource Filter: Before executing OnResourceExecuting on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Resource Filter: After executing OnResourceExecuting on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Action Filter: Before executing OnActionExecuting on filter Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Action Filter: After executing OnActionExecuting on filter Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Action Filter: Before executing OnActionExecuting on filter Microsoft.AspNetCore.Mvc.Infrastructure.ModelStateInvalidFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Action Filter: After executing OnActionExecuting on filter Microsoft.AspNetCore.Mvc.Infrastructure.ModelStateInvalidFilter.
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Executing action method [REMOVED].Controllers.SystemController.Get ([REMOVED]) - Validation state: Valid
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Executed action method [REMOVED].Controllers.SystemController.Get ([REMOVED]), returned result Microsoft.AspNetCore.Mvc.OkObjectResult in 2.0716ms.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Action Filter: Before executing OnActionExecuted on filter Microsoft.AspNetCore.Mvc.Infrastructure.ModelStateInvalidFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Action Filter: After executing OnActionExecuted on filter Microsoft.AspNetCore.Mvc.Infrastructure.ModelStateInvalidFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Action Filter: Before executing OnActionExecuted on filter Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Action Filter: After executing OnActionExecuted on filter Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Result Filter: Before executing OnResultExecuting on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Result Filter: After executing OnResultExecuting on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[4]
      Before executing action result Microsoft.AspNetCore.Mvc.OkObjectResult.
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.DefaultOutputFormatterSelector[11]
      List of registered output formatters, in the following order: Microsoft.AspNetCore.Mvc.Formatters.HttpNoContentOutputFormatter, Microsoft.AspNetCore.Mvc.Formatters.StringOutputFormatter, Microsoft.AspNetCore.Mvc.Formatters.StreamOutputFormatter, Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.DefaultOutputFormatterSelector[4]
      No information found on request to perform content negotiation.
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.DefaultOutputFormatterSelector[8]
      Attempting to select an output formatter without using a content type as no explicit content types were specified for the response.
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.DefaultOutputFormatterSelector[10]
      Attempting to select the first formatter in the output formatters list which can write the result.
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.DefaultOutputFormatterSelector[2]
      Selected output formatter 'Microsoft.AspNetCore.Mvc.Formatters.StringOutputFormatter' and content type 'text/plain' to write the response.
info: Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor[1]
      Executing ObjectResult, writing value of type 'System.String'.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[5]
      After executing action result Microsoft.AspNetCore.Mvc.OkObjectResult.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Result Filter: Before executing OnResultExecuted on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Result Filter: After executing OnResultExecuted on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Resource Filter: Before executing OnResourceExecuted on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
trce: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[3]
      Resource Filter: After executing OnResourceExecuted on filter Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter.
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Executed action [REMOVED].Controllers.SystemController.Get ([REMOVED]) in 716.9667ms
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 1713.0414ms 200 text/plain; charset=utf-8
```

For the non-local bad repro I'm not able to get onto the server at the moment so had to extract the logs from Azure instead so it looks a little different and it also means that I'm calling the endpoint through a gateway, but the end result is the same:

In chronological order:


```
2019-08-11T12:41:39.6937778Z, 
"Request starting HTTP/1.1 GET   https://[REMOVED]/SomeBasePath/api/system",
"{
""CategoryName"":""Microsoft.AspNetCore.Hosting.Internal.WebHost"",
""RequestPath"":""/SomeBasePath/api/system"",
""RequestId"":""800987b3-0000-d300-b63f84710c7967bb"",
""EventId"":""1"",
""Protocol"":""HTTP/1.1"",
""Method"":""GET"",
""Scheme"":""https"",
""Path"":""/SomeBasePath/api/system"",
""Host"":""[REMOVED]""
}"
```

```
2019-08-11T12:41:39.7003628Z,
"All hosts are allowed.",
"{"
"{OriginalFormat}"":""All   hosts are   allowed."",
""CategoryName"":""Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware"",
""RequestPath"":""/SomeBasePath/api/system"",
""RequestId"":""800987b3-0000-d300-b63f-84710c7967bb"
"}"
```

```
2019-08-11T12:41:39.7056025Z,
"Request did not match any routes",
"{
""{OriginalFormat}"":""Request did not match any   routes"",
""CategoryName"":""Microsoft.AspNetCore.Builder.RouterMiddleware"",
""RequestPath"":""/SomeBasePath/api/system"",
""RequestId"":""800987b3-0000-d300-b63f-84710c7967bb"",
""EventId"":""1""
}"
```

```
2019-08-11T12:41:39.975277Z,
"Request   finished in 281.5258ms 404   application/json",
{
""CategoryName"":""Microsoft.AspNetCore.Hosting.Internal.WebHost"",
""RequestPath"":""/SomeBasePath/api/system"",
""ElapsedMilliseconds"":""281.5258"",
""RequestId"":""800987b3-0000-d300-b63f-84710c7967bb"",
""EventId"":""2"",
""StatusCode"":""404"",
""ContentType"":""application/json""
}"
```

So, returns 404 but hits my `app.Run` terminal middleware so the message returned is 
`{ "message": "Not found!!!" }`

**Comment by Tratcher** [comment]
`Listening on prefix: http://localhost:5000/` and `Now listening on: http://localhost:5000/` should both have included the `/SomeBasePath` path base if it was defined at the server level.

Was this specified in the server for this run?
`options.UrlPrefixes.Add("http://localhost:5000/SomeBasePath/")`?
Or were you using UsePathBase for this run?

**Comment by dannyfhalpotia** [comment]
I was using UsePathBase for this run. Adding it to the UrlPrefix, does as you say add the "/SomeBasePath" to the two log statements. But the behaviour seems to be the same. It works on my own machine and not on the server.

**Comment by dannyfhalpotia** [comment]
@Tratcher 
I commented out `app.UsePathBase` and changed my UrlPrefix to this:

```
                .UseHttpSys(options =>
                {
                    options.Authentication.AllowAnonymous = true;
                    if (!env.IsDevelopment())
                    {
                        options.UrlPrefixes.Add("https://*:30310/SomeBasePath/");
                    }
                    else
                    {
                        options.UrlPrefixes.Add("http://localhost:5000/SomeBasePath/");
                    }
                })
```
Deploying the above, produces the logs below (same 404 result though). One noticeable difference is the presence of PathBase in the first log statement, which wasn't included when I was using app.UsePathBase() to set the base path. So it seems that it extracts this from the request such that we have a RequestPath without the base path...but yeah still getting the "Request did not match any routes". 

```
2019-08-11T17:52:38.0285455Z,
"Request starting HTTP/1.1 GET https://[REMOVED]/SomeBasePath/api/system",
"{
""CategoryName"":""Microsoft.AspNetCore.Hosting.Internal.WebHost"",
""RequestPath"":""/api/system"",
""RequestId"":""8001ddcc-0001-e000-b63f-84710c7967bb"",
""EventId"":""1"",
""Protocol"":""HTTP/1.1"",
""Method"":""GET"",
""Scheme"":""https"",
""Path"":""/api/system"",
""Host"":""[REMOVED]"",
""PathBase"":""/SomeBasePath""
}"
```

```
2019-08-11T17:52:38.0371785Z,
"All   hosts are   allowed.",
"{
""CategoryName"":""Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware"",
""{OriginalFormat}"":""All hosts are   allowed."",
""RequestPath"":""/api/system"",
""RequestId"":""8001ddcc-0001-e000-b63f-84710c7967bb""
}"
```

```
2019-08-11T17:52:38.0433558Z,
"Request did not match any routes",
"{
""CategoryName"":""Microsoft.AspNetCore.Builder.RouterMiddleware"",
""{OriginalFormat}"":""Request did not match any   routes"",
""RequestPath"":""/api/system"",
""RequestId"":""8001ddcc-0001-e000-b63f-84710c7967bb"",
""EventId"":""1""
}"
```

```
2019-08-11T17:52:38.2503602Z,
"Request finished in 221.803ms 404   application/json",
{
""CategoryName"":""Microsoft.AspNetCore.Hosting.Internal.WebHost"",
""ElapsedMilliseconds"":""221.803"",
""RequestPath"":""/api/system"",
""RequestId"":""8001ddcc-0001-e000-b63f-84710c7967bb"",
""StatusCode"":""404"",
""EventId"":""2"",
""ContentType"":""application/json""
}"
```

**Comment by Tratcher** [comment]
Yes the urlacl url parameter needs to exactly match what you pass into UrlPrefixes, including the path base (with a trailing slash).
`netsh http add urlacl url=https://*:30310/SomeBasePath/ user="NT AUTHORITY\NETWORK SERVICE"`

I'm starting to wonder if pathbase is really your main problem. It seems like you're also having a problem with your controllers being discovered for a published application. If you publish the app to a local folder and try to run it from there do you have the same issues?

**Comment by dannyfhalpotia** [comment]
I found the issue. Your second paragraph had me try out a few different things and changing this line in ConfigureServices did the trick:

`services.AddMvc()` to `services.AddMvc().AddApplicationPart(typeof(SystemController).Assembly);`

As I have my controllers defined in a separate dll, it seems that I have to specify the above in the startup.cs to let the MVC part know that it should look for controllers in another assembly. Should have added that little detail to my original question!

What is still a bit puzzling to me is how it managed to work when I ran it locally and only see this error when publishing to a server. That's why I was focussing so much on the paths etc.
